File: src/test/java/completion/CompletionTest.java
package completion;

import utils.TestUtilsKt;
import com.google.common.collect.Lists;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import javaproject.JavaProject;
import javax.swing.text.BadLocationException;
import javax.swing.text.Document;
import org.jetbrains.kotlin.completion.CompletionUtilsKt;
import org.netbeans.api.project.Project;
import org.netbeans.junit.NbTestCase;
import org.openide.filesystems.FileObject;
import org.openide.util.Exceptions;
import static junit.framework.TestCase.assertNotNull;
import org.jetbrains.kotlin.builder.KotlinPsiManager;
import org.jetbrains.kotlin.diagnostics.netbeans.parser.KotlinParser;
import org.jetbrains.kotlin.psi.KtFile;
import org.jetbrains.kotlin.resolve.AnalysisResultWithProvider;
import org.jetbrains.kotlin.resolve.KotlinAnalyzer;
import org.jetbrains.kotlin.utils.ProjectUtils;
import org.junit.Test;
import org.netbeans.modules.csl.api.CompletionProposal;

/**
 *
 * @author Alexander.Baratynski
 */
public class CompletionTest extends NbTestCase {

    private final Project project;
    private final FileObject completionDir;

    public CompletionTest() {
        super("Completion test");
        project = JavaProject.INSTANCE.getJavaProject();
        completionDir = project.getProjectDirectory().
                getFileObject("src").getFileObject("completion");
    }

    private void doTest(String fileName, Collection<String> items) {
        Document doc = TestUtilsKt.getDocumentForFileObject(completionDir, fileName);
        Collection<CompletionProposal> completionItems = null;
        List<CharSequence> completions = new ArrayList<CharSequence>();

        Integer caret = TestUtilsKt.getCaret(doc);
        assertNotNull(caret);

        FileObject file = ProjectUtils.getFileObjectForDocument(doc);
        KtFile ktFile = null;
        try {
            ktFile = KotlinPsiManager.INSTANCE.getParsedFile(file);

            AnalysisResultWithProvider resultWithProvider
                    = KotlinAnalyzer.INSTANCE.analyzeFile(project, ktFile);
            KotlinParser.setAnalysisResult(ktFile, resultWithProvider);
            
            completionItems = CompletionUtilsKt.createProposals(doc,
                    caret, resultWithProvider, "");
        } catch (IOException ex) {
            Exceptions.printStackTrace(ex);
        }

        assertNotNull(completionItems);

        for (CompletionProposal it : completionItems) {
            completions.add(it.getSortText());
        }

        boolean contains = false;
        if (completions.containsAll(items)) {
            contains = true;
        }

        assertEquals(true, contains);
    }

    @Test
    public void testProjectCreation() {
        assertNotNull(project);
        assertNotNull(completionDir);
    }

    @Test
    public void testStringCompletion() {
        doTest("checkStringCompletion.kt", Lists.newArrayList("toString()"));
    }

    @Test
    public void testBasicInt() {
        doTest("checkBasicInt.kt", Lists.newArrayList("Int"));
    }

    @Test
    public void testBasicAny() {
        doTest("checkBasicAny.kt", Lists.newArrayList("Any"));
    }

    @Test
    public void testAutoCastAfterIf() {
        doTest("checkAutoCastAfterIf.kt", Lists.newArrayList("value"));
    }

    @Test
    public void testAutoCastAfterIfMethod() {
        doTest("checkAutoCastAfterIfMethod.kt", Lists.newArrayList("test()"));
    }

    @Test
    public void testAutoCastForThis() {
        doTest("checkAutoCastForThis.kt", Lists.newArrayList("destroy()"));
    }

    @Test
    public void testAutoCastInWhen() {
        doTest("checkAutoCastInWhen.kt", Lists.newArrayList("left", "right"));
    }

    @Test
    public void testCompletionBeforeDotInCall() {
        doTest("checkCompletionBeforeDotInCall.kt",
                Lists.newArrayList("TestSample", "testVar", "testTop()", "testFun()"));
    }

    @Test
    public void testLocalLambda() {
        doTest("checkLocalLambda.kt", Lists.newArrayList("test()"));
    }

    @Test
    public void testCompanion() {
        doTest("checkCompanion.kt", Lists.newArrayList("companionVal", "companionFun()"));
    }

    @Test
    public void testExtendsClass() {
        doTest("checkExtendClass.kt", Lists.newArrayList("MyFirstClass", "MySecondClass"));
    }

    @Test
    public void testImport() {
        doTest("checkImport.kt", Lists.newArrayList("Proxy"));
    }

    @Test
    public void testInCallExpression() {
        doTest("checkInCallExpression.kt", Lists.newArrayList("func()"));
    }

    @Test
    public void testInClassInit() {
        doTest("checkInClassInit.kt", Lists.newArrayList("valExternal", "valInternal"));
    }

    @Test
    public void testInClassPropertyAccessor() {
        doTest("checkInClassPropertyAccessor.kt", Lists.newArrayList("test", "testParam"));
    }

    @Test
    public void testInImport() {
        doTest("checkInImport.kt", Lists.newArrayList("Proxy", "Base"));
    }

    @Test
    public void testInParameterType() {
        doTest("checkInParameterType.kt", Lists.newArrayList("Int"));
    }

}

File: src/test/java/completion/CompletionTest.kt
package completion

import utils.*
import javaproject.JavaProject
import javax.swing.text.Document
import org.jetbrains.kotlin.completion.*
import org.netbeans.api.project.Project
import org.netbeans.junit.NbTestCase
import org.openide.filesystems.FileObject
import org.jetbrains.kotlin.builder.KotlinPsiManager
import org.jetbrains.kotlin.diagnostics.netbeans.parser.KotlinParser
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.resolve.AnalysisResultWithProvider
import org.jetbrains.kotlin.resolve.KotlinAnalyzer
import org.jetbrains.kotlin.utils.ProjectUtils
import org.netbeans.modules.csl.api.CompletionProposal

/**
 *
 * @author Alexander.Baratynski
 */
class CompletionTest : NbTestCase("Completion test") {
    private val project: Project
    private val completionDir: FileObject

    init {
        project = JavaProject.javaProject
        completionDir = project.projectDirectory.getFileObject("src").getFileObject("completion")
    }

    private fun doTest(fileName: String, items: Collection<String>) {
        val doc = getDocumentForFileObject(completionDir, fileName)
        val caret = getCaret(doc)
        assertNotNull(caret)
        
        val file = ProjectUtils.getFileObjectForDocument(doc)
        val ktFile = KotlinPsiManager.getParsedFile(file)!!
        
        val resultWithProvider = KotlinAnalyzer.analyzeFile(project, ktFile)
        KotlinParser.setAnalysisResult(ktFile, resultWithProvider)
        
        val completionItems = createProposals(doc, caret, resultWithProvider, "")
        assertNotNull(completionItems)
        
        val completions = completionItems.map { it.sortText }
       
        assertEquals(true, completions.containsAll(items))
    }

    fun testProjectCreation() {
        assertNotNull(project)
        assertNotNull(completionDir)
    }

    fun testStringCompletion() = doTest("checkStringCompletion.kt", listOf("toString()"))

    fun testBasicInt() = doTest("checkBasicInt.kt", listOf("Int"))

    fun testBasicAny() = doTest("checkBasicAny.kt", listOf("Any"))

    fun testAutoCastAfterIf() = doTest("checkAutoCastAfterIf.kt", listOf("value"))
    
    fun testAutoCastAfterIfMethod() = doTest("checkAutoCastAfterIfMethod.kt", listOf("test()"))

    fun testAutoCastForThis() = doTest("checkAutoCastForThis.kt", listOf("destroy()"))

    fun testAutoCastInWhen() = doTest("checkAutoCastInWhen.kt", listOf("left", "right"))

    fun testCompletionBeforeDotInCall() = doTest("checkCompletionBeforeDotInCall.kt", listOf("TestSample", "testVar", "testTop()", "testFun()"))
    
    fun testLocalLambda() = doTest("checkLocalLambda.kt", listOf("test()"))

    fun testCompanion() = doTest("checkCompanion.kt", listOf("companionVal", "companionFun()"))

    fun testExtendsClass() = doTest("checkExtendClass.kt", listOf("MyFirstClass", "MySecondClass"))

    fun testImport() = doTest("checkImport.kt", listOf("Proxy"))

    fun testInCallExpression() = doTest("checkInCallExpression.kt", listOf("func()"))

    fun testInClassInit() = doTest("checkInClassInit.kt", listOf("valExternal", "valInternal"))

    fun testInClassPropertyAccessor() = doTest("checkInClassPropertyAccessor.kt", listOf("test", "testParam"))

    fun testInImport() = doTest("checkInImport.kt", listOf("Proxy", "Base"))

    fun testInParameterType() = doTest("checkInParameterType.kt", listOf("Int"))

    fun testUpperAndLowerCase() = doTest("checkUpperAndLowerCases.kt", listOf("method()"))
    
}